### DISCUSSION TITLE
CSV-Patch und Import von Bestandsdaten

### DISCUSSION THREAD

**Discussion by AmigaFreak** [Larian77/Openbiblio#38] [Q&A] [answered]
Hallo allerseits,

ich brauche dringend Hilfe mit OpenBiblio 0.8, welches ich kürzlich auf dem Webspace unseres lokalen Geschichtsvereins installiert habe, in dessen Vorstand ich aktiv bin.

Ich möchte den Bestand unserer Vereinsbibliothek importieren, der bisher in einer Word-Tabelle verwaltet wurde. Ich habe schnell gesehen habe daß es mir wahrscheinlich nicht in annehmbarer Zeit gelingen wird diese Daten ins MARC21-Format zu konvertieren. 
Daher bin ich auf den CSV-Patch gestoßen, welcher vor vielen Jahren mal auf Sourceforge veröffentlicht wurde. 
Nämlich diesen hier: https://sourceforge.net/p/obiblio/patches/_discuss/thread/e672eec9/4c4e/attachment/upload_csv.files-0.7.X-3.zip

Den habe ich installiert und er erscheint auch in meiner GUI. Aber offenbar funktioniert er nicht richtig mit Version 0.8 von OpenBiblio.
Soweit ich herausfinden konnte wird der PHP-Code in "upload_csv.php" zwischen Zeile 175 und 457 nicht richtig ausgeführt. Wenn man diesen auskommentiert, dann wird der Code danach ausgeführt, ansonsten nicht (ich habe übrigens keine Ahnung von PHP, aber ich bin Softwareentwickler in einem anderen Bereich; habe mich da also halt einfach etwas durchgewurstelt per Trial&Error). 

Nun ist die Frage: Kann mir hier eventuell jemand helfen dieses Problem zu finden und zu lösen? Generell wäre dieser CSV-Import ja eine sehr nützliche Erweiterung von OpenBiblio. Oder wie kann ich sonst meinen Bestand mit vertretbarem Aufwand importieren?

Schönen Gruß,
AmigaFreak

**Comment by LeAchim59** [comment]
@AmigaFreak 
Hallo,
diesen csv-Upload kenne ich gar nicht, Ich habe ein Programm das heißt openbiblioimport. Ich weiß nicht mehr, woher ich es habe, finde es auch im Internet nicht mehr. Das muss man im Ordner von openBiblio installieren und dann die import.php direkt aufrufen. Das Programm verlangt die Datenbankparameter als Eingabe und verarbeitet dann eine csv-Datei. 
Die "Originalversion" habe ich wohl auch nicht mehr, meine Version habe ich erweitert, dass noch einige Brettspielspezifische Felder eingelesen werden und ab Spalte x dann die Barcodenummern der Exemplare stehen, die dann auch direkt angelegt werden. Ansonsten werden die Grundinfos wie Titel, Autor, Standord, Genre, etc importiert. Könnte ich ggf. zur Verfügung stellen, muss aber für eigene Zwecke (Bücher) auf jeden Fall angepasst werden.

Eine andere Möglichkeit besteht darin, die Datenbanktabellen mit phpmyadmin zu bearbeiten. Da kann man auch csv-Dateien direkt importieren.

**Comment by AmigaFreak** [comment, marked as answer]
Hallo,
ich antworte mal in drei Teilen :-)

1. Ja, als Backuplösung würde dein "openbiblioimport" sicher in Betracht kommen. Wäre nett wenn du mir das irgendwie zu kommen lassen könntest.

2. Eines der Probleme bei dem verlinkten CSV-Addon scheint auf jeden Fall in Zeile 275 der upload_csv.php zu liegen. Das konnte ich herausfinden indem ich Schritt für Schritt Code auskommentiert habe. Da steht folgende if-Abfrage: if (!$copy->validateData()) {...}
Und bei $copy->validateData() "knallt" es offenbar. $copy ist eine Instanz der Klasse BiblioCopy. Warum das aber hier schief geht weiß ich nicht. Auch noch nicht ob noch weitere Fehler existieren, außer diesem in Zeile 275.
Irgendeine Idee warum das schief geht?

3. Folgendes soll nicht "böse" gemeint sein, aber hat sich denn bisher nie jemand Gedanken gemacht wie eine "kleine" Bibliothek ihren Bestand ohne viel Aufwand importieren kann? 
Das wäre doch ein Grundfunktion für so ein Projekt. 
Denn nach allem was ich bisher sehen kann ist MARC halt doch recht komplex und eine Konversion in dieses Format nicht mal eben aus dem Ärmel zu schütteln. Daher wäre sowas simples wie CSV doch sehr naheliegend. Mich wundert es halt daß es eine "einfache" Import-Möglichkeit im OpenBiblio bisher nicht gib.
PS.: Stimmt natürlich nicht ganz, denn irgendjemand hat ja um 2008 mal dieses CSV-Addon erstellt. Aber offenbar wurde das nicht weiter verfolgt.

**EDIT:
Wenn man die in Punkt 2. erwähnte if-Abfrage einfach auskommentiert und den Rest dann laufen lässt, funktioniert der CSV-Import anscheinend problemlos. Ich konnte so auf jeden Fall gerade tatsächlich zwei Bücher zu meiner OpenBiblio-Datenbank hinzufügen. Ich würde euch ja glatt (hinsichtlich Punkt 3) empfehlen dieses CSV-Addon zu übernehmen.**

**Comment by LeAchim59** [comment]
@AmigaFreak 
zu 1) Kriegen wir hin. Ich habe einen Fork zu openbiblio, da kann ich den Code reinstellen.

zu 2) Das validateData von BiblioCopy erwartet wohl einen Parameter. Es prüft aber nur ob ein Barcode vorhanden ist, kann man ggf. weglassen und vertrauen ;-)

zu 3) Das habe ich mich auch gefragt, als wir vor einigen Jahren anfingen und unsere Spieledatenbank erstmal importieren wollte. Ich bin dann eben auf das openbiblioimport gestoßen und habe es mir angepasst. Das Addon oder auch openbiblioimport zu übernehmen ist dann letztendlich nicht so trivial. Da nicht jeder unbedingt alle Felder benutzt, müsste man die Kopfzeile der csv-Datei interpretieren und jeweils die richtige Tabellenspalte zuordnen (alternativ bräuchte man ein csv-Datei mit ziemlich vielen Spalten. Und für Exemplare eigentlich eine zweite csv-Datei, in der man sich aber auf bibid beziehen muss, das aber vor dem Import der Medien nicht bekannt ist. Ist dann schon etwas aufwendiger und anscheinend hat das früher kaum jemand vermisst oder hat sich anders beholfen (zB eben über Datenbank-Tools wie phpMyAdmin, das ja auch importieren kann).

**Reply by AmigaFreak** [reply]
Also wie gesagt (der fettgedruckte Nachtrag in meinem vorherigen Posting) habe ich das Problem jetzt tatsächlich gelöst indem ich den validateData-Teil auskommentiert habe, wie du auch vorgeschlagen hast. 

> Da nicht jeder unbedingt alle Felder benutzt, müsste man die Kopfzeile der csv-Datei interpretieren und jeweils die richtige Tabellenspalte zuordnen (alternativ bräuchte man ein csv-Datei mit ziemlich vielen Spalten.

Ich bin jetzt nicht ganz sicher was du meinst, aber das CSV-Addon hat eine variable Kopfzeile. Es gibt nur drei "mandatory" Spalten (Titel, Autor und Call Number) und einige optionale Spalten. Zusätzlich kann man beliebig viele MARC-Daten als Spalte hinzufügen, indem man einfach das zugehörige MARC-Tag in die Kopfzeile einfügt.

Allerdings kann man tatsächlich keine Exemplare erstellen, soweit ich das sehe.

**Comment by Larian77** [comment]
Hi,

Ich habe das Projekt ja auch nur übernommen. Im Originalprojekt gab es diese Funktion nicht und ich selber habe diese nie benötigt, da ich meine Datenbanken eh neu aufgesetzt habe.
Grundsätzlich ist das auf alle Fälle eine ser Funktionen in meiner Liste, die dringend ins Hauptprojekt mit rein müssen :-)

**Reply by AmigaFreak** [reply]
Ja, wie gesagt funktioniert dieses CSV-Addon eigentlich recht gut, wenn man den weiter oben erwähnten Code-Teil auskommentiert. Ihr könnt ja mal drüber nachdenken ob ihr das fixen und dann übernehmen wollt. 

Eine Frage hab ich aber noch: Ich bin auf meiner Suche nach einer Möglichkeit des CSV-Imports auf folgende Version von OpenBiblio (inklusive ausführlicher Dokumentation) gestoßen: 
https://openbiblio.github.io/openbiblio_docs/root.html
Dort wird diese Version als OpenBiblio 1.0 bezeichnet und auch der CSV-Import ist dort beschrieben.
**Ist das ein verwaister Fork?** Denn Eure Version hier ist ja 0.8.
Wenn man so ganz naiv erstmals auf OpenBiblio stößt, dann ist die Versionsgeschichte etwas verwirrend.

**Reply by LeAchim59** [reply]
Zur Geschichte kann Dir am ehesten @Larian77 was erzählen. 
Wir haben die letzte aktive Version 0.7.3 als Ausgangsbasis gewählt, die musste dringend auf PHP8 aktualisiert werden, da älteres PHP von Webhostern nicht mehr gerne gesehen wird. Und das nehmen wir jetzt auch als Ausgangsbasis für weitere Entwicklungen, weil niemand genau weiß, was in dieser Verion 1.0 alles gemacht wurde und welchen Stand das alles hat.

**Comment by superudo** [comment]
Die Frage, die sich hier immer wieder stellt: Wie soll das Importformat aussehen, damit es zum einen leicht genug zu bedienen ist und andererseits komplex genug ist, um alle relevanten Daten zu übertragen. CSV ist zwar recht weit verbreitet, ist aber leider nicht wirklich standardisiert, so dass jedes Programm seine eigene Version von CSV schreibt, JSON ist zur Zeit "in", aber welches Programm exportiert seine Daten schon in dem Format. Da bleiben also noch viele Fragen offen. Je nach Anwendung sind es auch gern unterschiedliche Medien, Metadaten und andere Daten, die übertragen werden sollen. Wird wohl auf den "Kleinsten gemeinsamen Nenner" hinauslaufen... 

Da ich selbst gern die Daten unseres bisherigen Leihsystems nach Openbiblio übertragen möchte, wird das sicher ein Punkt sein, zu dem ich beitragen werde.
